### PR TITLE
Add CnotRing template

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**

Adds a new template to the library, which is a very basic architecture intended as a "vanilla" example: rotations on each qubit are followed by a "ring" (closed nearest neighbour chain) of CNOTs.

This template is easier to use and understand than ``StronglyEntanglingLayers`` which we tend to use in demonstrations a lot. 

cnot_ring

**Description of the Change:**

* add template file
* import template in module
* update ``introduction/templates.rst``
* add template tests
* add integration tests
* add init function for parameters
* add init tests

**Benefits:**

Growing the template library.

### CODE REVIEW

**Comment by mariaschuld** [comment]
When I run ``make coverage`` locally, the template module is still covered at 100%.

**Comment by mariaschuld** [comment]
Note: I added a border to the other layer icon images, so there are some non-essential binary image files changed in this PR. :)

**Comment by mariaschuld** [comment]
Happy to discuss a better name :)

File: pennylane/init.py [hunk 1]
```diff
@@ -835,3 +835,57 @@ def interferometer_varphi_normal(n_wires, mean=0, std=0.1, seed=None):
...
+
+    params = np.random.normal(loc=mean, scale=std, size=(n_layers, n_wires))
+    return params
+
+
+def cnot_ring_layers_uniform(n_layers, n_wires, low=0, high=2 * pi, seed=None):
+    r"""Creates a parameter array for :func:`~.CnotRingLayers`, drawn from a uniform
+    distribution.
+
+    The shape of the parameter array is ``(n_layers, n_wires)`` and each parameter is drawn uniformly at random \
```
Comment (antalszava) [inline pennylane/init.py]:
```suggestion
    The shape of the parameter array is ``(n_layers, n_wires)`` and each parameter is drawn uniformly at random 
```

File: pennylane/templates/layers/cnot_ring.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+
+@template
+def CnotRingLayers(weights, wires, rotation=None):
+    r"""Layers consisting of one-parameter single-qubit rotations on each qubit, followed by a closed chain
+    or *ring* of CNOT gates.
+
+    The ring of CNOT gates connects every qubit with its neighbour,
+    whereas the last qubit is considered to be a neighbour of the first qubit.
```
Comment (antalszava) [inline pennylane/templates/layers/cnot_ring.py]: (Feel free to leave it as is if you prefer the current wording, just thought to leave this as a suggestion )
```suggestion
    with the last qubit being considered as a neighbour to the first qubit.
```

File: pennylane/templates/layers/cnot_ring.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    .. figure:: ../../_static/templates/layers/cnot_ring.png
+        :align: center
+        :width: 40%
+        :target: javascript:void(0);
+
+    The argument ``weights`` is an array of weights for each layer. The number of layers :math:`L` is derived
+    from the first dimension of ``weights``. When using a single wire, the template only applies the single
```
Comment (antalszava) [inline pennylane/templates/layers/cnot_ring.py]: Merged the content of the first sentence into the second in a brief way, perhaps it could also just simply be removed (details of `weights` may be found later in the **Parameters** part)
```suggestion
    The number of layers :math:`L` is determined by the first dimension of the argument ``weights``. When using a single wire, the template only applies the single
```

File: pennylane/templates/layers/cnot_ring.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        In this case, only one CNOT gate is applied in each layer:
+
+        .. figure:: ../../_static/templates/layers/cnot_ring_2wires.png
+            :align: center
+            :width: 30%
+            :target: javascript:void(0);
+
+    Args:
+
+        weights (array[float]): array of weights of shape ``(L, len(wires))``
```
Comment (antalszava) [inline pennylane/templates/layers/cnot_ring.py]:
```suggestion
        weights (array[float]): array of weights with shape ``(L, len(wires))``, each weight is used as a parameter for the rotation
```

File: pennylane/templates/layers/cnot_ring.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+            n_wires = 3
+            dev = qml.device('default.qubit', wires=n_wires)
+
+            @qml.qnode(dev)
+            def circuit(weights):
+                CnotRingLayers(weights=weights, wires=range(n_wires))
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_wires)]
+
+        >>> circuit([[pi, pi, pi]])
```
Comment (antalszava) [inline pennylane/templates/layers/cnot_ring.py]: Would it make things unintuitive if in this case (`n_layers=1`), we could also accept a list instead of a nested list?

This would require a change in `_check_shape` from what I see (I can imagine users trying - myself included - passing `[pi, pi, pi]` here for the first time):
```python
/pennylane/pennylane/templates/utils.py in _check_shape(inpt, target_shape, msg, bound)
    121     else:
    122         if shape != target_shape:
--> 123             raise ValueError(msg)
    124 
    125     return shape

ValueError: 'weights' must be of shape (3, 3); got (3,)
```

Comment (josh146) [reply pennylane/templates/layers/cnot_ring.py]: Worth thinking about, but I find it more intuitive with the weight shape remaining consistent

File: pennylane/templates/layers/cnot_ring.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                CnotRingLayers(weights=weights, wires=range(n_wires))
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_wires)]
+
+        >>> circuit([[pi, pi]])
+        [-1, 1]
+
+
+        **Changing the rotation gate**
+
+        Another single-qubit gate than ``RX`` can be used, as long as it only takes a single parameter.
```
Comment (antalszava) [inline pennylane/templates/layers/cnot_ring.py]:
```suggestion
        Any single-qubit gate can be used as a rotation gate, as long as it only takes a single parameter. The default is the ``RX`` gate.
```

File: tests/test_templates.py [hunk 7]
```diff
@@ -534,7 +539,19 @@ def make_n_features(self, n):
                   (QAOAEmbedding,
                    {'features': [1., 2.], 'weights': qaoa_embedding_normal(n_layers=2, n_wires=1), 'wires': range(1)}),
                   (QAOAEmbedding,
-                   {'features': [1., 2.], 'weights': qaoa_embedding_uniform(n_layers=2, n_wires=1), 'wires': range(1)})
+                   {'features': [1., 2.], 'weights': qaoa_embedding_uniform(n_layers=2, n_wires=1), 'wires': range(1)}),
+                  (CnotRingLayers,
+                   {'weights': cnot_ring_layers_uniform(n_layers=3, n_wires=1), 'wires': range(1)}),
```
Comment (antalszava) [inline tests/test_templates.py]: Although easy to check, `n_layers=1` would be one of the simplest edge cases. I see that this holds for other templates as well, though I guess if it's included it could be included simply for this template only.

File: tests/test_templates_layers.py [hunk 8]
```diff
@@ -462,3 +462,75 @@ def test_random_layer_weights(self, n_subsystems, tol):
...
+class TestCnotRing:
+    """Tests for the CnotRingLayers method from the pennylane.templates.layers module."""
+
+    @pytest.mark.parametrize("n_wires, n_cnots", [(1, 0),
+                                                  (2, 1),
+                                                  (3, 3),
+                                                  (4, 4)])
+    def test_circuit_queue(self, n_wires, n_cnots):
+        """Tests the gate types and parameter values in the circuit."""
+        np.random.seed(12)
```
Comment (antalszava) [inline tests/test_templates_layers.py]: How come seeding is used here?

File: pennylane/templates/layers/cnot_ring.py (hunk 5 above)
Comment (mariaschuld) [reply pennylane/templates/layers/cnot_ring.py]: Me too, we may risk ambiguities.

File: tests/test_templates_layers.py (hunk 8 above)
Comment (mariaschuld) [reply tests/test_templates_layers.py]: I think that's a copy and paste error, thanks!

File: tests/test_templates.py (hunk 7 above)
Comment (mariaschuld) [reply tests/test_templates.py]: I added a test case of ``n_layers=1, n_wires=1``. Was that what you had in mind?

Comment (antalszava) [reply tests/test_templates.py]: Sure, sounds good to me :)

**Comment by mariaschuld** [comment]
Thanks @antalszava for your comments. I implemented them all...

**Comment by mariaschuld** [comment]
Actually, I'm also unhappy with the name - this is an issue for all templates.

There are two general strategies: 
1. name it after the gates/structures used (in which case we could consider adding the ``Rot`` part).  
2. name after papers (in this case there is none)
3. brand the templates with new names

I am almost tending towards 3, and make sure we make ample use of the images/icons of templates.

Then, it would be great to have a very simple, basic name for this one. Like ``BasicEntangler``?

**Comment by josh146** [comment]
> I am almost tending towards 3, and make sure we make ample use of the images/icons of templates.
>
> Then, it would be great to have a very simple, basic name for this one. Like BasicEntangler?

I concur!

**Comment by antalszava** [comment]
`BasicEntangler` sounds amazing to me as well! :)

File: pennylane/templates/layers/basic_entangler.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        weights (array[float]): array of weights with shape ``(L, len(wires))``, each weight is used as a parameter for the rotation
+        wires (Sequence[int] or int): qubit indices that the template acts on
+        rotation (pennylane.ops.Operation): one-parameter single-qubit gate to use,
+                                            if ``None``, :class:`~pennylane.ops.RX` is used as default
+    Raises:
+        ValueError: if inputs do not have the correct format
+
+    .. UsageDetails::
+
+        The template is used inside a qnode:
```
Comment (antalszava) [inline pennylane/templates/layers/basic_entangler.py]:
```suggestion
        The template is used inside a :class:`~.BaseQNode`:
```

File: tests/test_templates_layers.py [hunk 10]
```diff
@@ -462,3 +462,74 @@ def test_random_layer_weights(self, n_subsystems, tol):
...
+
+
+class TestCnotRing:
+    """Tests for the BasicEntanglerLayers method from the pennylane.templates.layers module."""
+
+    @pytest.mark.parametrize("n_wires, n_cnots", [(1, 0),
+                                                  (2, 1),
+                                                  (3, 3),
+                                                  (4, 4)])
+    def test_circuit_queue(self, n_wires, n_cnots):
```
Comment (antalszava) [inline tests/test_templates_layers.py]: Separating this into two tests: `test_circuit_queue_types` and `test_circuit_queue_parameter_values`, could make any errors arising be more explicit.

**Comment by antalszava** [comment]
Two suggestions (a minor and a bit bigger change) came to my mind during the other review, happy without the test reorganization too.

File: pennylane/templates/layers/basic_entangler.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        expected_shape,
+        msg="'weights' must be of shape {}; got {}" "".format(expected_shape, _get_shape(weights)),
+    )
+
+    ###############
+
+    for layer in range(repeat):
+
+        broadcast(unitary=rotation, pattern="single", wires=wires, parameters=weights[layer])
+        broadcast(unitary=CNOT, pattern="ring", wires=wires)
```
Comment (thisac) [inline pennylane/templates/layers/basic_entangler.py]: Still being fairly new to PennyLane, I find this quite awesome. A very compact, nice and clear way of writing these circuits. 😄 👍

File: pennylane/templates/layers/basic_entangler.py (hunk 9 above)
Comment (josh146) [reply pennylane/templates/layers/basic_entangler.py]: Question: do we want to link to `BaseQNode` here, or the `QNode` constructor? The second is more userfacing, and the docstring makes more sense for users, even thought the first is more accurate.

File: pennylane/templates/layers/basic_entangler.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    _check_shape,
+    _check_no_variable,
+    _check_wires,
+    _check_number_of_layers,
+    _get_shape,
+)
+
+
+@template
+def BasicEntanglerLayers(weights, wires, rotation=None):
```
Comment (josh146) [inline pennylane/templates/layers/basic_entangler.py]: Very annoying comment here (😆) but: what are your thoughts on naming this `BasicEntangler` rather than `BasicEntanglerLayers`?

e.g.,
```python
@qml.qnode(dev)
def circuit(params):
    qml.templates.BasicEntagler(params, wires=[0, 1])
    return [qml.PauliZ(i) for i in range(5)]
```

File: pennylane/templates/layers/basic_entangler.py (hunk 9 above)
Comment (antalszava) [reply pennylane/templates/layers/basic_entangler.py]: To be honest, I was also not sure about this. I'm happy if the `QNode` constructor is referenced at these places. One thing that would be nice to update is the following link (to `BaseQNode` I guess): https://github.com/XanaduAI/pennylane/blob/a76f103695e15668c22005a86a6d8e31d9b3c236/pennylane/qnodes/decorator.py#L35

Users could then quickly surf through this function, be directed to `BaseQNode` find the other variety of qnodes as well.

This change of the link would be out of scope for sure. :)

File: pennylane/templates/layers/basic_entangler.py (hunk 12 above)
Comment (antalszava) [reply pennylane/templates/layers/basic_entangler.py]: (Agreed! Or if `Layers` stays in the name, `BasicEntanglingLayers` to match `StronglyEntanglingLayers`? )

Comment (josh146) [reply pennylane/templates/layers/basic_entangler.py]: I woudn't even mind `StronglyEntangler`!

File: pennylane/templates/layers/basic_entangler.py (hunk 9 above)
Comment (josh146) [reply pennylane/templates/layers/basic_entangler.py]: Agree, the constructor should not be linking to itself in the docstring! But perhaps also add a warning in the BaseQNode (and other QNodes), that the are for advanced users, and the constructor should generally be used.

**Comment by mariaschuld** [comment]
Need to overwrite codecov, locally the templates module is still 100% tested:

Screenshot from 2020-03-24 09-12-00
